### DISCUSSION TITLE
arrow function vs function declaration in Next.js

### DISCUSSION THREAD

**Discussion by JuhwanLeeKR** [vercel/next.js#37835] [Help] [answered]
It may be a very basic question because I have just studied next.js, so please answer...

I used Arrow function to construct components in the react.js project, but I think next.js usually uses function declaration

Is this because of the SSR? Or does it not matter if I use both? I'd appreciate it if you could explain...

**Comment by icyJoseph** [comment, marked as answer]
For Next.js and React, it virtually does not matter at all.

The main different between arrow functions and old functions is that, old functions have a `this` object bound to them, and the `this` object has mechanics that most developers nowadays are not very familiar with, and which are not used as much in React development, but for example, a backend framework like Fastify does use `this` sometimes.

There's also hoisting, which means you can use something before it is defined, but since we work with modules, this is seldom a problem.

```js
console.log(foo)
function foo(){}
```

vs

```js
console.log(foo) // fails, tried to use before initialisation
const foo = () => {} 
```

There's also a practical difference, IE11 doesn't understand arrow function syntax, which means that, when serving IE11, arrow functions code has to be transpiled to functions. 

Other than that, besides some other stuff with generators, and TypeScript ergonomics, that's about it, you can use them interchangeably at this point.

**Reply by JuhwanLeeKR** [reply]
I'm getting help again Thank you so much!

**Reply by jooncco** [reply]
Another (seldom problematic) case when you need old function declaration is when you use `fs` inside `getStaticProps`.  
I tried to use `fs` inside `getStaticProps` arrow function, and it failed to find Module `fs` at compile time.  
After I changed this into old function declaration, it worked like a charm.

**Reply by GOLDER303** [reply]
Could you please elaborate on what you mean by TypeScript ergonomics?

**Reply by icyJoseph** [reply]
Well, to begin with assertion functions, which cannot be used with arrow functions:

- https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#assertion-functions
- https://github.com/microsoft/TypeScript/issues/34523

And also that when you assign a function to a value, you can type the entire function. Though this not exclusively of arrow functions:

```js
type Foo = VoidFunction
const foo:Foo = () => {}
const bar:Foo = function (){}
```

Also function overloading requires a bit of interface magic:

```ts
interface Overloaded {
    (x:number): string[];
    (x:boolean): number[];
}

const foo: Overloaded =  (x: any): any => {
    if (typeof x === 'boolean') return [];

    if (typeof x === 'number') return [];

    throw new Error('x')
}

function bar(x:number):string
function bar(x:string):number
function bar(x:any): any {
    if (typeof x === 'string') return 0;

    if (typeof x === 'number') return '1';

    throw new Error('x')
}
```

As you can see, these are not really a big deal.

**Reply by GOLDER303** [reply]
Thank you very much for the explanation

**Comment by atwright147** [comment]
Does it make any difference in API Routes?

**Reply by icyJoseph** [reply]
Not really. In Next.js apps, the `this` keyword is not typically used by developers, nor the framework. If you use something like Fastify, for example, then you need to use functions to allow for context binding. Since API routes in Next.js share no implicit context, you can use either.
